File: src/web_app/modern_web/src/lib/domain/BeatData.ts
/**
 * Beat Domain Model
 * 
 * Immutable data structure for individual beats in kinetic sequences.
 * Based on the modern desktop app's BeatData adapted for TypeScript.
 */

import type { PictographData } from './PictographData';
import type { MotionData } from './MotionData';

export interface BeatData {
  readonly id: string;
  readonly beat_number: number;
  readonly duration: number;
  readonly blue_reversal: boolean;
  readonly red_reversal: boolean;
  readonly is_blank: boolean;
  readonly pictograph_data?: PictographData | null;
  readonly metadata: Record<string, any>;
}

export function createBeatData(data: Partial<BeatData> = {}): BeatData {
  return {
    id: data.id ?? crypto.randomUUID(),
    beat_number: data.beat_number ?? 1,
    duration: data.duration ?? 1.0,
    blue_reversal: data.blue_reversal ?? false,
    red_reversal: data.red_reversal ?? false,
    is_blank: data.is_blank ?? false,
    pictograph_data: data.pictograph_data ?? null,
    metadata: data.metadata ?? {},
  };
}

export function updateBeatData(beat: BeatData, updates: Partial<BeatData>): BeatData {
  return {
    ...beat,
    ...updates,
  };
}

export function isValidBeat(beat: BeatData): boolean {
  return beat.duration >= 0 && beat.beat_number >= 0;
}

export function getBeatLetter(beat: BeatData): string | undefined {
  return beat.pictograph_data?.letter ?? beat.metadata.letter;
}

export function hasPictograph(beat: BeatData): boolean {
  return beat.pictograph_data != null;
}

export function getBlueMotion(beat: BeatData): MotionData | null {
  if (beat.pictograph_data?.motions) {
    return beat.pictograph_data.motions.blue ?? null;
  }
  return null;
}

export function getRedMotion(beat: BeatData): MotionData | null {
  if (beat.pictograph_data?.motions) {
    return beat.pictograph_data.motions.red ?? null;
  }
  return null;
}

export function createBeatFromPictograph(
  pictograph_data: PictographData, 
  beat_number: number
): BeatData {
  return createBeatData({
    beat_number,
    pictograph_data,
    metadata: {
      letter: pictograph_data.letter,
      created_from_pictograph: true,
    },
  });
}

export function beatDataToObject(beat: BeatData): Record<string, any> {
  return {
    id: beat.id,
    beat_number: beat.beat_number,
    duration: beat.duration,
    blue_reversal: beat.blue_reversal,
    red_reversal: beat.red_reversal,
    is_blank: beat.is_blank,
    pictograph_data: beat.pictograph_data,
    metadata: beat.metadata,
  };
}

export function beatDataFromObject(data: Record<string, any>): BeatData {
  return createBeatData({
    id: data.id,
    beat_number: data.beat_number,
    duration: data.duration,
    blue_reversal: data.blue_reversal,
    red_reversal: data.red_reversal,
    is_blank: data.is_blank,
    pictograph_data: data.pictograph_data,
    metadata: data.metadata,
  });
}

File: src/desktop/modern/domain/models/beat_data.py
"""
Beat Domain Models

Immutable data structures for individual beats in kinetic sequences.
Handles beat data, motion references, and glyph information.
"""

from __future__ import annotations

from dataclasses import dataclass, field
import json

# Forward reference for PictographData to avoid circular imports
from typing import TYPE_CHECKING, Any
import uuid

from .motion_data import MotionData


if TYPE_CHECKING:
    from .pictograph_data import PictographData


@dataclass(frozen=True)
class BeatData:
    """
    REPLACES: Beat class with UI coupling

    Pure business data for a single beat in a sequence.
    No UI dependencies, completely immutable.
    """

    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    beat_number: int = 1
    duration: float = 1.0
    blue_reversal: bool = False
    red_reversal: bool = False
    is_blank: bool = False

    # NEW: Optional pictograph data
    pictograph_data: PictographData | None = None

    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        """Validate beat data."""
        if self.duration < 0:
            raise ValueError("Duration must be positive")
        if self.beat_number < 0:
            raise ValueError("Beat number must be non-negative")

    def update(self, **kwargs) -> BeatData:
        """Create a new BeatData with updated fields."""
        from dataclasses import replace

        return replace(self, **kwargs)

    def is_valid(self) -> bool:
        """Check if beat has valid data for sequence inclusion."""
        return True

    @property
    def has_pictograph(self) -> bool:
        """Check if this beat has associated pictograph data."""
        return self.pictograph_data is not None

    @property
    def letter(self) -> str | None:
        """Get beat letter from pictograph data if available, fallback to metadata."""
        if self.has_pictograph and self.pictograph_data.letter:
            return self.pictograph_data.letter
        return self.metadata.get("letter")

    @property
    def blue_motion(self) -> MotionData | None:
        """Get blue motion from pictograph data if available."""
        if self.has_pictograph and self.pictograph_data.motions:
            return self.pictograph_data.motions.get("blue")
        return None

    @property
    def red_motion(self) -> MotionData | None:
        """Get red motion from pictograph data if available."""
        if self.has_pictograph and self.pictograph_data.motions:
            return self.pictograph_data.motions.get("red")
        return None

    @classmethod
    def from_pictograph(
        cls, pictograph_data: PictographData, beat_number: int
    ) -> BeatData:
        """Create BeatData from PictographData."""
        return cls(
            beat_number=beat_number,
            pictograph_data=pictograph_data,
            metadata={
                "letter": pictograph_data.letter,
                "created_from_pictograph": True,
            },
        )

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for serialization."""
        result = {
            "id": self.id,
            "beat_number": self.beat_number,
            "duration": self.duration,
            "blue_reversal": self.blue_reversal,
            "red_reversal": self.red_reversal,
            "is_blank": self.is_blank,
            "metadata": self.metadata,
        }

        # Add pictograph data if present
        if self.pictograph_data:
            result["pictograph_data"] = self.pictograph_data.to_dict()

        return result

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> BeatData:
        """Create from dictionary."""
        # Handle pictograph data
        pictograph_data = None
        if "pictograph_data" in data:
            from .pictograph_data import PictographData

            pictograph_data = PictographData.from_dict(data["pictograph_data"])

        return cls(
            id=data.get("id", str(uuid.uuid4())),
            beat_number=data.get("beat_number", 1),
            duration=data.get("duration", 1.0),
            blue_reversal=data.get("blue_reversal", False),
            red_reversal=data.get("red_reversal", False),
            is_blank=data.get("is_blank", False),
            pictograph_data=pictograph_data,
            metadata=data.get("metadata", {}),
        )

    def to_camel_dict(self) -> dict[str, Any]:
        """Convert to dictionary with camelCase keys for JSON APIs."""
        from ..serialization import dataclass_to_camel_dict

        return dataclass_to_camel_dict(self)

    def to_json(self, camel_case: bool = True, **kwargs) -> str:
        """Serialize to JSON string."""
        from ..serialization import domain_model_to_json

        if camel_case:
            return domain_model_to_json(self, **kwargs)
        return json.dumps(self.to_dict(), **kwargs)

    @classmethod
    def from_json(cls, json_str: str, camel_case: bool = True) -> BeatData:
        """Create instance from JSON string."""
        from ..serialization import domain_model_from_json

        if camel_case:
            return domain_model_from_json(json_str, cls)
        data = json.loads(json_str)
        return cls.from_dict(data)

    @classmethod
    def empty(cls) -> BeatData:
        """Create an empty beat data."""
        return cls(is_blank=True)
